materialize-databricks: use stateKey for tracking binding state

williamhbaker opened the pull request:
**Description:**

_Related PR: https://github.com/estuary/connectors/pull/1213. This was merged and all active databricks materializations are now running on it and emitting post-ack zero'd checkpoints._

Adds support for state keys to materialize-databricks. The backfill capability more-or-less worked even without this, but the full `stateKey` implementation mitigates certain unlikely edge cases that may occur if the ack state is not persisted, and a table is re-backfilled and then persisted queries for the prior table are run against the re-created table.

I manually tested migrating an existing materialization from the prior image over to this, as well as starting a new materialization from scratch with this image. Also various combinations of multiple bindings with or without data and disabling & re-enabling them to make sure everything looked right.

See issue [#1146](https://github.com/estuary/connectors/issues/1146)

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/connectors/1215)
<!-- Reviewable:end -->

williamhbaker commented on materialize-databricks/driver.go:
```diff
@@ -426,14 +445,12 @@ func (d *transactor) Store(it *m.StoreIterator) (_ m.StartCommitFunc, err error)
 
-		converted, err := b.target.ConvertAll(it.Key, it.Values, it.RawJSON)
-		if err != nil {
+		if err := b.storeFile.start(ctx); err != nil {
```
Added the `err` check from `storeFile.start` here

mdibaiee commented:
@williamhbaker I think we need to have the same logic as https://github.com/estuary/connectors/pull/1218 here

williamhbaker commented:
> @williamhbaker I think we need to have the same logic as #1218 here

:+1: I'll update this after that one merges.

williamhbaker commented on materialize-snowflake/snowflake.go:
```diff
@@ -528,9 +528,6 @@ func (d *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error
 	// the queries are actually executed and done
 	var results = make(map[string]stdsql.Result)
 	for stateKey, item := range d.cp {
-		if len(item.Query) == 0 {
```
Since we are only ever fully deleting a key/value pair from `d.cp` or adding an item with a non-zero length query, I'm not adding this check to `materialize-databricks` and removing it from snowflake for consistency.

williamhbaker commented:
@mdibaiee I rebased after https://github.com/estuary/connectors/pull/1218 merged and updated this implementation to be as consistent with `materialize-snowflake` as I could - let me know what you think.
